RecoLocalTracker/SiStripClusterizer/plugins/SiStripRawToClusterGPUKernel.cu

```diff
@@ -0,0 +1,557 @@
+#include <cub/cub.cuh>
+
+#include "HeterogeneousCore/CUDAUtilities/interface/allocate_device.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/allocate_host.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/currentDevice.h"
+
+#include "HeterogeneousCore/CUDAUtilities/interface/host_unique_ptr.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/device_unique_ptr.h"
+
+#include "HeterogeneousCore/CUDAUtilities/interface/copyAsync.h"
+
+#include "CUDADataFormats/SiStripCluster/interface/SiStripClustersCUDA.h"
+
+#include "CalibFormats/SiStripObjects/interface/SiStripClusterizerConditionsGPU.h"
+
+#include "ChannelLocsGPU.h"
+#include "SiStripRawToClusterGPUKernel.h"
+#include "StripDataView.cuh"
+
+//#define GPU_DEBUG
+#if defined(EDM_ML_DEBUG) || defined(GPU_DEBUG)
+#define GPU_CHECK
+#include <stdio.h>
+#endif
+
+namespace stripgpu {
+  using ConditionsDeviceView = SiStripClusterizerConditionsGPU::Data::DeviceView;
+
+  __global__ static void unpackChannels(const ChannelLocsView *chanlocs,
+                                        const ConditionsDeviceView *conditions,
+                                        uint8_t *alldata,
+                                        uint16_t *channel,
+                                        stripgpu::stripId_t *stripId) {
+    const int tid = threadIdx.x;
+    const int bid = blockIdx.x;
+    const int nthreads = blockDim.x;
+
+    const auto chan = nthreads * bid + tid;
+    if (chan < chanlocs->size()) {
```

Comment:
I would suggest to change the `if` into a `for` loop, to let the kernel run with an arbitrary number of blocks and threads:
```suggestion
    const auto first = nthreads * bid + tid;
    const auto stride = blockDim.x * gridDim.x;
    for (auto chan = first; chan < chanlocs->size(); chan += stride) {
```

Reply:
@fwyzard , why this `if` condition should be equivalent to the `for` loop? Just for my education

Reply:
CUDA (and any GPU runtime I know of) organise the work in blocks (or groups) of threads (or elements).

The number of blocks is given by `gridDim.x` (_i.e._ the size of the kernel launch _grid_ along the _x_ dimension).
The number of threads per block is given by `blockDim.x` (_i.e._ the size of the kernel launch _blocks_ along the _x_ dimension).
So the total number of threads is `blockDim.x * gridDim.x`.

When this is larger than the problem size (in this case, `chanlocs->size()`), there are enough blocks and threads to process all the elements of the problem, and an `if` is used to limit the execution to the problem size and avoid an overflow.

However, it _might_ be more efficient to use a smaller number of blocks, and reuse the threads to process multiple elements.
The extreme case of a single block with a single element has very bad performance, but is extremely useful for debugging :-)

Due to the way the GPU memory fetching works, it is more efficient to make a _strided loop_ where a thread works over its own work element, then one offset by the grid size (or the block size, but that's more complicated to set up).

So, in the end, the advice we give is to replace an `if` condition like
```c++
auto thread = blockDim.x * blockIdx.x + threadIdx.x;
if (thread < size) {
  ...
}
```
with a `for` loop like
auto first = blockDim.x * blockIdx.x + threadIdx.x;
auto stride = blockDim.x * gridDim.x;
for (auto thread = first; thread < size; thread += stride) {
  ...
}
```